wip: Pkg3-style package loading

StefanKarpinski opened the pull request:
This works for `julia --compiled-modules=no` but fails in the presence of module precompilation, e.g. with the following output:
```jl
julia> import SHA
77394: find_package(Main, SHA)
77394: find_package_in_project(/Users/stefan/projects/Pkg3.jl/Project.toml, SHA)
77394: find_package_in_manifest(/Users/stefan/projects/Pkg3.jl/Manifest.toml, ea8e919c-243c-51af-8825-aaa63cd721ce, IOStream(<file /Users/stefan/projects/Pkg3.jl/Manifest.toml>))
77394: find_package(Main, SHA)
77394: find_package_in_project(/Users/stefan/projects/Pkg3.jl/Project.toml, SHA)
77394: find_package_in_manifest(/Users/stefan/projects/Pkg3.jl/Manifest.toml, ea8e919c-243c-51af-8825-aaa63cd721ce, IOStream(<file /Users/stefan/projects/Pkg3.jl/Manifest.toml>))
[ Info: Precompiling module SHA @ Base loading.jl:960
77395: find_package(SHA, Compat)
77395: find_package_in_manifest(/Users/stefan/projects/Pkg3.jl/Manifest.toml, ea8e919c-243c-51af-8825-aaa63cd721ce, Compat)
77395: find_package_in_manifest(/Users/stefan/projects/Pkg3.jl/Manifest.toml, Compat, IOStream(<file /Users/stefan/projects/Pkg3.jl/Manifest.toml>))
77395: find_package(SHA, Compat)
77395: find_package_in_manifest(/Users/stefan/projects/Pkg3.jl/Manifest.toml, ea8e919c-243c-51af-8825-aaa63cd721ce, Compat)
77395: find_package_in_manifest(/Users/stefan/projects/Pkg3.jl/Manifest.toml, Compat, IOStream(<file /Users/stefan/projects/Pkg3.jl/Manifest.toml>))
77394: find_package(Main, Compat)
77394: find_package_in_project(/Users/stefan/projects/Pkg3.jl/Project.toml, Compat)
┌ Warning: Module `SHA` declares `__precompile__(true)` but `require` failed
│ to create a usable precompiled cache file
│   exception = ErrorException("Required dependency Compat not found in current path.")
└ @ Base loading.jl:774
ERROR: Module SHA declares __precompile__(true) but require failed to create a usable precompiled cache file.
Stacktrace:
 [1] error at ./error.jl:33 [inlined]
 [2] _require(::Module, ::Symbol) at ./loading.jl:777
 [3] require(::Module, ::Symbol) at ./loading.jl:626
```
This is from the `Pkg3` package since it's one of the only examples at this point of a project with Pkg3-style project and manifest files. @vtjnash: I think the ball is in your court here – I've tried to get this to work with module precompilation, but the logic of that code is beyond me. The fundamental problem seems to be that while `Compat` is loaded into `SHA` in the precompile child process and in the main process in the absence of module precompilation, when precompilation is on, the main process tries to load `Compat` into `Main` instead  of `SHA`, which is incorrect and leads to a load failure.

The other major todo is to change the root module key to being a UUID instead of a name when a module is resolved by UUID, but that can be done in a separate PR since it's strictly internal and non-breaking (it adds the ability to have multiple root modules loaded with the same name).

StefanKarpinski committed:
require(mod::Symbol) => require(into::Module, mod::Symbol)

Thread an `into` module through all require-related calls so that we
can look up the name of the module to be loaded in the context of
the module into which it is being loaded.

test/compile: comment out failing code

StefanKarpinski committed:
new Pkg3-style code loading working without precompilation

StefanKarpinski commented on base/loading.jl:
```diff
@@ -80,25 +80,331 @@ else
+        return new(bytes)
     end
 end
+SHA1(s::Union{String,SubString{String}}) = SHA1(hex2bytes(s))
```
This is just internal, we should not export this type or anything and should maybe replace it with something else, but for now it can live here and we can clean this stuff up later.

StefanKarpinski commented on base/initdefs.jl:
```diff
@@ -49,31 +49,83 @@ Determine whether Julia is running an interactive session.
-        prepend!(LOAD_PATH, split(ENV["JULIA_LOAD_PATH"], @static Sys.iswindows() ? ';' : ':'))
-    end
     push!(LOAD_PATH, abspath(BINDIR, "..", "local", "share", "julia", "site", vers))
     push!(LOAD_PATH, abspath(BINDIR, "..", "share", "julia", "site", vers))
```
This currently leaves the `site` paths in here and the second one is where `stdlib` lives, which we need. However, I think that we should probably have a `Project.toml` and `Manifest.toml` for Julia itself which includes these. The installed stdlib packages should probably be in a builtin/standard depot instead, but that's kind of a detail. We can always put relative paths into `stdlib` from the manifest file.

StefanKarpinski commented on base/loading.jl:
```diff
@@ -419,42 +744,50 @@ function _require(mod::Symbol)
+        if info isa Tuple{String,String,UUID}
+            # horrible hack, but since we don't control module creation, ¯\_(ツ)_/¯
+            ccall(:jl_set_global, Cvoid, (Any, Any, Any), __toplevel__, ENVINFO, info[2:end])
+        end
```
If we constructed the module object to be loaded and then evaluated code into it, then we could explicitly set the `ENVINFO` name in there instead of doing nasty stuff like this. I think we should probably do that, but I didn't want to couple this change with a change like that, so I just hacked around it instead.

StefanKarpinski commented on base/loading.jl:
```diff
@@ -471,26 +804,23 @@ include_string(m::Module, txt::String, fname::String) =
     include_string(m, String(txt), String(fname))
 
-function source_path(default::Union{AbstractString,Nothing}="")
+function tls_recurse(key::Symbol, default=nothing)
```
I was at some point trying passing ENVINFO via task local storage, which is why I refactored this, but I didn't end up using it. I'm leaving the refactoring in since it's good to separate the TLS recursion from source path lookup anyway.

vtjnash replied:
This method is deprecated (v0.7) anyways, I'm not sure it matters

StefanKarpinski commented on base/loading.jl:
```diff
@@ -564,6 +898,10 @@ function create_expr_cache(input::String, output::String, concrete_deps::typeof(
+                  const $ENVINFO = $envinfo
+                  # same horrible hack as in _require
+                  ccall(:jl_set_global, Cvoid, (Any, Any, Any),
+                        Base.__toplevel__, $(Meta.quot(ENVINFO)), $envinfo)
```
This allows correct new-style module loading in the child precompile process. The failure is later, however, when the precompiled module needs to get loaded into the original parent process, which tries to load it into Main.

StefanKarpinski commented on base/reflection.jl:
```diff
@@ -31,6 +33,30 @@ Base.LinAlg
+
+Get the root module enclosing `m`: this is the top-level module loaded by
+`import` or `using` which ultimately create `m`. This root module is found by
+iteratively taking a module's parent until self-parented module is found.
```
I didn't mean to leave this in the PR – it was something I wrote and then didn't end up using, but it's somewhat useful, so maybe we should leave it in.

StefanKarpinski commented on src/dump.c:
```diff
@@ -1927,7 +1927,7 @@ static int size_isgreater(const void *a, const void *b)
 }
 
-static jl_value_t *read_verify_mod_list(ios_t *s, arraylist_t *dependent_worlds, jl_array_t *mod_list)
+static jl_value_t *read_verify_mod_list(jl_module_t *from, ios_t *s, arraylist_t *dependent_worlds, jl_array_t *mod_list)
```
Oops: should this be called `from` or `into`? I'm not sure what this module is for. @JeffBezanson, @vtjnash?

vchuravy replied:
I think this is `into` since it should verify the module list with respect to the environment the module is loaded into.

vtjnash replied:
This isn't loading a module into an environment, it's just parsing a file and asserting that the header looks valid. @StefanKarpinski You should drop all changes to this file; my earlier PR to loading.jl moved everything into Julia that matters for Pkg3.

StefanKarpinski committed:
add Pkg.dir to LOAD_PATH (to keep Documenter working)

vchuravy commented on base/initdefs.jl:
```diff
@@ -49,31 +49,84 @@ Determine whether Julia is running an interactive session.
+end
+
 function init_load_path(BINDIR = Sys.BINDIR)
+    load_path = get(ENV, "JULIA_LOAD_PATH", "@|@v#.#.#|@v#.#|@v#|@default|@!v#.#")
```
For other readers see https://github.com/JuliaLang/Pkg3.jl/issues/82 for context.
The string `"@|@v#.#.#|@v#.#|@v#|@default|@!v#.#"` turns into:

```
LOAD_PATH = Any[
    [ CurrentProject(),
      NamedEnv(“v$(VERSION.major).$(VERSION.minor).$(VERSION.patch)”),
      NamedEnv(“v$(VERSION.major).$(VERSION.minor)”),
      NamedEnv(“v$(VERSION.major)”),
      NamedEnv(“default”),
      NamedEnv(“v$(VERSION.major).$(VERSION.minor)”, create=true),
]
```

StefanKarpinski replied:
Yes, we need to bikeshed this syntax, but this is what I ended up going with:

* `@` = current project environment found by looking in parent directories
* `@name` = named environment if it already exists
* `@!name` = named environment even if it doesn't already exist
* `#` characters are replaced by the major, minor and patch version numbers in order

I decided not to introduce any escaping since you can always use `.juliarc.jl` to set the values programmatically. If you have an actual relative path name that starts with a `@` you can always write it as `./@...`

vchuravy reviewed (commented):
I am still trying to wrap my head around the implications of the environment. As far as I understand it is now possible to have the name of a package resolve to different packages/versions depending on the environment (and therefore) module the package is being loaded into. This leads me to think that we will need a cache-file per environment.

In general this looks good, even though it would benefit from comments. This is my third read and I am still not sure that I understand environments and their implications fully.

StefanKarpinski commented on src/toplevel.c:
```diff
@@ -186,6 +190,7 @@ jl_value_t *jl_eval_module_expr(jl_module_t *parent_module, jl_expr_t *ex)
         newm->parent = parent_module;
         b->value = (jl_value_t*)newm;
         jl_gc_wb_binding(b, newm);
+        // copy parent environment into submodule
```
Ah, yes, that comment is a stray, left from a previous iteration where the above code was here.

StefanKarpinski commented:
> I am still trying to wrap my head around the implications of the environment. As far as I understand it is now possible to have the name of a package resolve to different packages/versions depending on the environment (and therefore) module the package is being loaded into. This leads me to think that we will need a cache-file per environment.

In essence, in an explicit environment the project file assigns identities to top-level package names via a mapping from names to package UUIDs in the `[deps]` section of the project file. When loading indirect dependencies via a manifest file, names are mapped to UUIDs via individual `[deps]` section for each package that is loaded. That means that `import A` can load one package at the top level and a different thing while loading a dependency. This is necessary to make sure that dependencies keep working no matter what other name clashes may occur in the ecosystem.

Caching absolutely should happen on a per-environment level. However, only the first environment in `LOAD_PATH` is guaranteed to be loaded intact – the following environments may end up getting dependencies at versions that appear in earlier environments in `LOAD_PATH`. So all of the caching is really specific to the first environment in that sense.

StefanKarpinski commented on base/sysimg.jl:
```diff
@@ -488,22 +492,22 @@ using Base
-Base.require(:Distributed)
-Base.require(:Printf)
-Base.require(:Future)
+Base.require(Base, :Base64)
```
They are, aren't they?

vtjnash replied:
yes (I'm assuming this does _not_ bind the name `Base.Base64`, right?)

vtjnash commented on base/loading.jl:
```diff
@@ -142,9 +453,10 @@ function _require_from_serialized(mod::Symbol, path_to_try::String, depmodnames:
-            modpath = find_package(string(modname))
-            modpath === nothing && return ErrorException("Required dependency $modname not found in current path.")
-            mod = _require_search_from_serialized(modname, String(modpath))
+            info = find_package(into, string(modname))
```
this (this thing that is being found) isn't being loaded into `into`, it's being loaded into `mod`

vtjnash replied:
but also `mod` and `modname` aren't Module names, they're UUIDs

StefanKarpinski replied:
Except they're not actual UUIDs, they're the 64-bit integers you're calling "UUIDs". Also, why is `modname` called `modname` if it's not a module name? Can you throw me a bone here and tell me how to get the module associated with `mod` instead of just telling me what's wrong with the code?

StefanKarpinski replied:
The `into` argument to `_require` is what's passed in here and as far as I can tell, it results in the correct module being passed in here. I think you may not be understanding what I'm trying to accomplish here. I've tried to explain and we've not gotten anywhere; this all works correctly in the absence of module precompilation, so if you could please make a RP on top of this one that fixes all the precompilation stuff, that would be great.

StefanKarpinski replied:
As far as I can tell, `mod` is the name of the module that needs to be created. `into` is the module into which the binding for `mod` is to be created when the process is done. So it seems to me that `mod` is **not** the module that should be passed as the `into` argument here. Why do you think it is?

StefanKarpinski commented on base/loading.jl:
```diff
@@ -80,25 +80,333 @@ else
+import Base.Random: UUID
+
+const SlugInt = UInt32 # max p = 4
+const chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
```
which is 78 characters versus 64.

Keno replied:
I agree that @vtjnash's way is more obviously free of errors.

StefanKarpinski replied:
I've already updated this to `String(['A':'Z'; 'a':'z'; '0':'9'])` locally, which is shorter and clearer than either.

StefanKarpinski committed:
loading: shorter more explicit definition of slug chars

StefanKarpinski commented:
Unrelated LAPACK failure and a timeout. The previous run was all green; this is just a rebase.

StefanKarpinski committed:
toplevel.c: move stray comment

vtjnash committed:
track requires dependencies alongside include dependencies

StefanKarpinski commented:
I don't get how this passed any tests. It fails locally since `loaded_modules` is used but not defined anywhere.

vtjnash commented:
None of them ran the tests (since it is not mergeable). Strange that they still post success.

StefanKarpinski committed:
Revert "require: refactor error handling"

This reverts commit f53209354b1c802ded472466ccf0635018d0c8e0.

StefanKarpinski committed:
loading: allow finding packages without actually loading code

also only store the UUID of each module, not the manifest path

StefanKarpinski committed:
loading: add test project

StefanKarpinski committed:
loading: rename #ENVINFO to #uuid since that's all it is now

StefanKarpinski committed:
find_package: always return (path, uuid) pair [ci skip]

If `path` is `nothing` the package wasn't found
If `uuid` is `nothing` the package has no UUID

StefanKarpinski commented:
Superseded by #25455.
